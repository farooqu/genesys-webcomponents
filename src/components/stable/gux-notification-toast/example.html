<div>
  <h1>gux-notification-toast</h1>
  <h2>Accents</h2>
  <gux-notification-toast accent="neutral">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is the message</div>
  </gux-notification-toast>

  <gux-notification-toast accent="positive">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is a message that is long but is one line</div>
  </gux-notification-toast>

  <gux-notification-toast accent="alert">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">
      <div>This is the first line</div>
      <div>This is the second line</div>
    </div>
  </gux-notification-toast>
  <gux-notification-toast accent="warning">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is the message</div>
  </gux-notification-toast>
</div>

<div class="stop-default">
  <h2>Dismiss event default prevented</h2>
  <gux-notification-toast accent="neutral">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is the message</div>
  </gux-notification-toast>
  <gux-notification-toast accent="neutral">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is the message</div>
  </gux-notification-toast>
  <gux-notification-toast accent="neutral">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is the message</div>
  </gux-notification-toast>
  <gux-notification-toast accent="neutral">
    <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
    <div slot="title">Title</div>
    <div slot="message">This is the message</div>
  </gux-notification-toast>
</div>

<style
  onload="(function() {
  const test = document.querySelector('.stop-default');
  test.addEventListener('guxdismiss', (event) => {
    event.preventDefault();

    event.target.classList.toggle('red');
  })
})()"
>
  .red {
    background-color: red !important;
  }
</style>
