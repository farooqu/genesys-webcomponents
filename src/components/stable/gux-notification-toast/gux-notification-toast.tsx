import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  Host,
  JSX,
  Prop
} from '@stencil/core';

import { trackComponent } from '../../../usage-tracking';

import { GuxNotificationToastAccent } from './gux-notification-toast.types';

/**
 * @slot icon - Required slot for gux-icon
 * @slot title - Required slot for the notification toast title
 * @slot message - Required slot for the notification toast message
 */
@Component({
  styleUrl: 'gux-notification-toast.less',
  tag: 'gux-notification-toast',
  shadow: true
})
export class GuxNotificationToast {
  /**
   * The component accent.
   */
  @Prop()
  accent: GuxNotificationToastAccent = 'neutral';

  @Event()
  guxdismiss: EventEmitter<void>;

  @Element()
  private root: HTMLElement;

  componentWillLoad(): void {
    trackComponent(this.root, { variant: this.accent });
  }

  render(): JSX.Element {
    return (
      <Host>
        <div class={`gux-icon gux-${this.accent}`}>
          <slot name="icon" />
        </div>

        <div class="gux-content">
          <div class="gux-title">
            <slot name="title" />
          </div>

          <div class="gux-message">
            <slot name="message" />
          </div>
        </div>

        <gux-dismiss-button
          onClick={this.onDismissClickHandler.bind(this)}
        ></gux-dismiss-button>
      </Host>
    ) as JSX.Element;
  }

  private onDismissClickHandler(event: MouseEvent): void {
    event.stopPropagation();

    const dismissEvent = this.guxdismiss.emit();
    if (!dismissEvent.defaultPrevented) {
      this.root.remove();
    }
  }
}
