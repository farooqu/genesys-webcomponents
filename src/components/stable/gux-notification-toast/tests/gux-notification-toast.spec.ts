import { newSpecPage } from '@stencil/core/testing';
import { GuxNotificationToast } from '../gux-notification-toast';

const components = [GuxNotificationToast];
const language = 'en';

describe('gux-notification-toast', () => {
  describe('#render', () => {
    [
      {
        description: 'should render neutral notification toast',
        html: `
          <gux-notification-toast accent="neutral">
            <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
            <div slot="title">Title</div>
            <div slot="message">This is the message</div>
          </gux-notification-toast>
        `
      },
      {
        description: 'should render positive notification toast',
        html: `
          <gux-notification-toast accent="positive">
            <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
            <div slot="title">Title</div>
            <div slot="message">This is the message</div>
          </gux-notification-toast>
        `
      },
      {
        description: 'should render alert notification toast',
        html: `
          <gux-notification-toast accent="alert">
            <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
            <div slot="title">Title</div>
            <div slot="message">This is the message</div>
          </gux-notification-toast>
        `
      },
      {
        description: 'should render warning notification toast',
        html: `
          <gux-notification-toast accent="warning">
            <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
            <div slot="title">Title</div>
            <div slot="message">This is the message</div>
          </gux-notification-toast>
        `
      }
    ].forEach(({ description, html }) => {
      it(description, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.rootInstance).toBeInstanceOf(GuxNotificationToast);

        expect(page.root).toMatchSnapshot();
      });
    });
  });

  describe('dismiss', () => {
    it('click dismiss button', async () => {
      const html = `
        <gux-notification-toast accent="neutral">
          <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
          <div slot="title">Title</div>
          <div slot="message">This is the message</div>
        </gux-notification-toast>
      `;
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const dismissButton =
        element.shadowRoot.querySelector('gux-dismiss-button');
      const guxdismissSpy = jest.fn();
      const clickSpy = jest.fn();
      const elementRemoveSpy = jest.spyOn(element, 'remove');

      page.win.addEventListener('guxdismiss', guxdismissSpy);
      page.win.addEventListener('click', clickSpy);

      dismissButton.click();
      await page.waitForChanges();

      expect(guxdismissSpy).toHaveBeenCalled();
      expect(clickSpy).not.toHaveBeenCalled();
      expect(elementRemoveSpy).toBeCalledWith();
    });

    it('click dismiss button and prevent default', async () => {
      const html = `
        <gux-notification-toast accent="neutral">
          <gux-icon slot="icon" icon-name="user-add" decorative></gux-icon>
          <div slot="title">Title</div>
          <div slot="message">This is the message</div>
        </gux-notification-toast>
      `;
      const page = await newSpecPage({ components, html, language });
      const element = page.root as HTMLElement;
      const dismissButton =
        element.shadowRoot.querySelector('gux-dismiss-button');
      const elementRemoveSpy = jest.spyOn(element, 'remove');

      page.win.addEventListener('guxdismiss', (event: Event) => {
        event.preventDefault();
      });

      dismissButton.click();
      await page.waitForChanges();

      expect(elementRemoveSpy).not.toBeCalled();
    });
  });
});
