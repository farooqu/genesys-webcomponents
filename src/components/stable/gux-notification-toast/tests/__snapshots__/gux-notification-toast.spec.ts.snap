// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-notification-toast #render should render alert notification toast 1`] = `
<gux-notification-toast accent="alert">
  <!---->
  <div class="gux-alert gux-icon">
    <gux-icon decorative="" icon-name="user-add" slot="icon"></gux-icon>
  </div>
  <div class="gux-content">
    <div class="gux-title">
      <div slot="title">
        Title
      </div>
    </div>
    <div class="gux-message">
      <div slot="message">
        This is the message
      </div>
    </div>
  </div>
  <button class="gux-dismiss-button" title="Dismiss">
    <gux-icon icon-name="close" screenreadertext="Dismiss"></gux-icon>
  </button>
</gux-notification-toast>
`;

exports[`gux-notification-toast #render should render neutral notification toast 1`] = `
<gux-notification-toast accent="neutral">
  <!---->
  <div class="gux-icon gux-neutral">
    <gux-icon decorative="" icon-name="user-add" slot="icon"></gux-icon>
  </div>
  <div class="gux-content">
    <div class="gux-title">
      <div slot="title">
        Title
      </div>
    </div>
    <div class="gux-message">
      <div slot="message">
        This is the message
      </div>
    </div>
  </div>
  <button class="gux-dismiss-button" title="Dismiss">
    <gux-icon icon-name="close" screenreadertext="Dismiss"></gux-icon>
  </button>
</gux-notification-toast>
`;

exports[`gux-notification-toast #render should render positive notification toast 1`] = `
<gux-notification-toast accent="positive">
  <!---->
  <div class="gux-icon gux-positive">
    <gux-icon decorative="" icon-name="user-add" slot="icon"></gux-icon>
  </div>
  <div class="gux-content">
    <div class="gux-title">
      <div slot="title">
        Title
      </div>
    </div>
    <div class="gux-message">
      <div slot="message">
        This is the message
      </div>
    </div>
  </div>
  <button class="gux-dismiss-button" title="Dismiss">
    <gux-icon icon-name="close" screenreadertext="Dismiss"></gux-icon>
  </button>
</gux-notification-toast>
`;

exports[`gux-notification-toast #render should render warning notification toast 1`] = `
<gux-notification-toast accent="warning">
  <!---->
  <div class="gux-icon gux-warning">
    <gux-icon decorative="" icon-name="user-add" slot="icon"></gux-icon>
  </div>
  <div class="gux-content">
    <div class="gux-title">
      <div slot="title">
        Title
      </div>
    </div>
    <div class="gux-message">
      <div slot="message">
        This is the message
      </div>
    </div>
  </div>
  <button class="gux-dismiss-button" title="Dismiss">
    <gux-icon icon-name="close" screenreadertext="Dismiss"></gux-icon>
  </button>
</gux-notification-toast>
`;
